persist: Write structured columnar data to S3

ParkMyCar opened the pull request:
This PR adds a V0 implementation of writing structured columnar data to S3.

To start it introduces two `dyncfg`s to control this:
1. `persist_batch_columnar_format` - controls the format we'll use to write batches, either `Row` which is the same as the current behavior, or `Both` which writes row data and structured columnar format.
2. `persist_part_decode_Format` - controls how we decode parts, current options are `Row { validate_structured: bool }`. We will always decode `Row` data, but we will optionally decode and validate the structured data.

Using these two `dyncfg`s we update the Arrow/Parquet handling to optionally encode and decode two new `k_s` and `v_s` columns which are structured versions of the existing `k` and `v` columns.

If structured encoding is turned on, we use the columnar `Part` we create for collecting statistics and instead of throwing it away we write it to S3. So the write path shouldn't become any more expensive. If structured decoding is enabled then we will optionally decode a `K` and `V` from our structured data, check that it matches what we decode from the Row data, and record the result of that check in a prometheus metric for monitoring.

#### Tests

I extended some existing tests that validate round-tripping `ScalarType`s/`Datum`s to go through Parquet. Also updated the `testdrive` mzcompose framework to support setting dyncfgs and addded a new nightly test slug that runs all of testdrive with structured data enabled.

#### Metrics

Added a few new metrics to track the encoding speed and size of the structured format.

Arrow related metrics

(for `k_s` and `v_s` columns)
* `decode_count`
* `decode_seconds`
* `correct_count`
* `incorrect_count`

(for a `Part`)
* `part_build_count`
* `part_build_seconds`

Parquet Related metrics

(for `k`, `v`, `t`, `d`, `k_s`, `v_s` columns)
* `uncompressed_size`
* `compressed_size`

(for a Parquet file)
* `encoded_size`
* `num_row_groups`

### Motivation

Progress towards https://github.com/MaterializeInc/materialize/issues/24830

### Tips for reviewer

This PR is split up into logical commits that can be reviewed independently:

1. Changes to our existing Columnar code that is necessary to support round-tripping through Parquet.
2. Add relevant Arrow and Parquet methods for encoding the new `k_s` and `v_s` structured columns.
3. Changes to the `Batch` encoding path to optionally write structured data
4. Changes to the `Batch` decoding path to optionally read structured data
5. Test related changes.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  <!-- Reference the design in the description. -->
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  <!-- Ask in #team-cloud on Slack if you need help preparing the cloud PR. -->
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Adds a feature gated ability to write structured columnar data.

bkirwi commented on src/persist/src/indexed/columnar.rs:
```diff
@@ -523,6 +524,22 @@ impl ColumnarRecords {
+    pub key: Option<StructArray>,
+    /// The structured `v` column.
+    pub val: Option<StructArray>,
+}
```
Are there any circumstances in which one of these should be set and the other is not? If not, thoughts on moving the option up one level?

ParkMyCar replied:
AFAIU [`arrow2` doesn't allow empty struct arrays](https://github.com/MaterializeInc/materialize/blob/main/src/persist-types/src/part.rs#L63-L67), so if `key` or `val` are ever empty (which `val` almost always is) it gets modeled as `None` here.

bkirwi commented on src/persist/src/indexed/columnar/parquet.rs:
```diff
@@ -124,25 +137,57 @@ fn encode_parquet_kvtd<W: Write>(
+            Ok(BlobTraceUpdates::Row(ret))
+        }
+        Some(ProtoFormatMetadata::StructuredMigration(1)) => {
+            // It's an invariant that we only write a single Row Group.
```
OOC, why change this as part of the migration? (If we've ever hashed this out together I've forgotten the details!)

ParkMyCar replied:
We didn't hash this out, and is something I need to add to the design doc!

Our current stats architecture (i.e. `PartStats`) represents an entire part. If we have more than one Row Group then we'd end up generating multiple `PartStats` that we would need to merge together. I made this change actually but when chatting with @danhhz we realized having a single Row Group is also nice for projection pushdown.

AFAIU [a single Parquet file is broken into Row Groups](https://parquet.apache.org/docs/concepts/), and each Row Group is broken into Column chunks. If we had a table with columns 'a', 'b', and 'c', and a parquet file with multiple Row Groups, then to fetch column 'a' we would need to issue multiple GET requests to S3 to get the 'a' Column Chunks in each Row Group. By enforcing a single Row Group it means we need only a single GET request to fetch a column.

I'm [pretty sure we already write a single Row Group](https://github.com/MaterializeInc/materialize/blob/main/src/persist-client/src/internal/encoding.rs#L1374-L1382) but formally enforcing it seems like a good idea. Also, switching from writing a single Row Group to multiple Row Groups seems much easier to do than the reverse!

bkirwi replied:
Understood, thanks!

bkirwi commented on test/testdrive/mzcompose.py:
```diff
@@ -55,6 +55,13 @@ def workflow_default(c: Composition, parser: WorkflowArgumentParser) -> None:
+        type=str,
+        action="append",
+        nargs="*",
+        help="Dyncfgs to enable in Materialize",
```
This looks like it would support any system parameter, so I'd be tempted to name it more generally!

ParkMyCar replied:
Good call, updated!

bkirwi reviewed (commented):
Nice! I'm going to need to chew on this a bit before having any deep feedback - leaving a first round of more surface / nitpicky comments in the meantime.

ParkMyCar commented on src/persist-client/src/batch.rs:
```diff
@@ -318,6 +318,13 @@ pub(crate) const BATCH_DELETE_ENABLED: Config<bool> = Config::new(
+pub(crate) const BATCH_COLUMNAR_FORMAT: Config<&'static str> = Config::new(
+    "persist_batch_columnar_format",
+    BatchColumnarFormat::default().as_str(),
+    "Columnar format for a batch written to Persist (Materialize).",
```
Updated!

bkirwi commented on src/persist-client/src/fetch.rs:
```diff
@@ -44,6 +48,30 @@ use crate::internal::state::{BatchPart, HollowBatchPart};
+pub const PART_DECODE_FORMAT: Config<&'static str> = Config::new(
+    "persist_part_decode_format",
+    PartDecodeFormat::default().as_str(),
+    "Format we'll use to decode a Persist 'Part' (Materialize).",
```
Same comment as the other config - it would be helpful to mention the valid values here.

ParkMyCar replied:
Added

bkirwi reviewed (approved):
Done a second pass!

Overall feel good about it. A lot of moving parts, but that's par for this sort of migration, and we'll have a chance to tighten things up later on.

Happy to see the testdrive run, but we should make sure to ping testing team, and it'd be pretty interesting to see how nightly behaves with this fully enabled before merging.

ParkMyCar commented on src/persist/src/indexed/columnar/arrow.rs:
```diff
@@ -105,6 +119,62 @@ pub fn encode_arrow_batch_kvtd(x: &ColumnarRecords) -> Chunk<Box<dyn Array>> {
+
+    /// Generate a list of encodings for all of the "leaf" arrays.
+    ///
+    /// TODO(parkmycar): It's probably worth it to use encodings other than "Plan".
```
Updated!

ParkMyCar commented on src/persist/src/indexed/columnar/arrow.rs:
```diff
@@ -121,6 +191,7 @@ pub(crate) const ENABLE_ARROW_LGALLOC_NONCC_SIZES: Config<bool> = Config::new(
 pub fn decode_arrow_batch_kvtd(
     x: &Chunk<Box<dyn Array>>,
     metrics: &ColumnarMetrics,
+    expected_cols: usize,
```
Ah ha! I like that a lot better, refactored to pass around `&[Box<dyn Array>]`s

ParkMyCar commented on src/persist/src/indexed/columnar/parquet.rs:
```diff
@@ -124,25 +137,57 @@ fn encode_parquet_kvtd<W: Write>(
+            let batch = reader
+                .next()
+                .ok_or_else(|| Error::String("found empty batch".to_string()))?;
+            assert!(reader.next().is_none());
```
We did check the row group size above this, but it was a bit disjointed. I refactored this to central the error checking

ParkMyCar commented on src/persist-client/src/batch.rs:
```diff
@@ -1002,22 +1005,54 @@ impl<T: Timestamp + Codec64> BatchParts<T> {
+                {
+                    let columnar_start = Instant::now();
+                    let columnar_result = encode_updates(&schemas, &updates.updates);
+                    let _columnar_duration = columnar_start.elapsed();
```
removed! In a follow-up PR I'll add tracking for how long it takes to columnar encode

ParkMyCar commented on src/persist-client/src/fetch.rs:
```diff
@@ -721,6 +787,37 @@ where
             });
+
+            // Note: We only provide structured columns, if they were originally written, and a
+            // dyncfg was specified to run validation.
```
I tried a little bit to refactor like you mentioned, but couldn't come up with a super clean solution. I agree it is repetitive though and will keep this in the back of my mind as I do more work here

ParkMyCar commented:
> Happy to see the testdrive run, but we should make sure to ping testing team, and it'd be pretty interesting to see how nightly behaves with this fully enabled before merging.

The testdrive run currently fails actually, there is a panic inside of `arrow2` that I need to dig into, not sure if it's our usage or a bug in the crate. Regardless I will also reach out to the testing team to get additional coverage here.

danhhz commented on src/persist-client/src/batch.rs:
```diff
@@ -318,6 +318,13 @@ pub(crate) const BATCH_DELETE_ENABLED: Config<bool> = Config::new(
     "Whether to actually delete blobs when batch delete is called (Materialize).",
 );
 
+/// The columnar format we write down a batch with.
```
nit: if it's pub crate, don't need the rustdoc (the dyncfg description seems sufficient IMO)

ParkMyCar replied:
updated!

danhhz commented on src/persist-client/src/fetch.rs:
```diff
@@ -44,6 +46,30 @@ use crate::internal::state::{BatchPart, HollowBatchPart};
 use crate::read::LeasedReaderId;
 use crate::ShardId;
 
+/// The columnar format we write down a batch with.
```
nit: copy-pasta on the rustdoc? seems like we could make it pub(crate) and then remove the rustdoc

ParkMyCar replied:
done

danhhz commented on src/persist-client/src/fetch.rs:
```diff
@@ -1004,6 +1039,63 @@ impl<T: Timestamp + Codec64> RustType<(ProtoLeasedBatchPart, Arc<Metrics>)> for
+            },
+            x => {
+                let default = PartDecodeFormat::default();
+                tracing::error!("Invalid part decode format: '{x}', falling back to {default}");
```
we should do something that fails CI here

ParkMyCar replied:
done!

danhhz commented on src/persist/src/indexed/encoding.rs:
```diff
@@ -26,14 +26,71 @@ use timely::progress::{Antichain, Timestamp};
+            "both" => BatchColumnarFormat::Both,
+            x => {
+                let default = BatchColumnarFormat::default();
+                tracing::error!("Invalid batch columnar type: {x}, falling back to {default}");
```
ditto should do something that fails CI here

ParkMyCar replied:
Started using `soft_panic_or_log!` macro which does exactly that

danhhz commented on src/persist/src/indexed/encoding.rs:
```diff
@@ -151,7 +151,61 @@ pub struct BlobTraceBatchPart<T> {
+        }
+    }
+
+    /// Returns the number of updates.
```
is this doc correct? looks like it returns the number of `ColumnarRecords` chunks, which each could have any number of updates

edit: oh, I see. the name of this method is confusing IMO. `num_row_groups` or something?

ParkMyCar replied:
Updated the name to be `num_row_groups`

danhhz commented on src/persist-client/src/stats.rs:
```diff
@@ -121,16 +121,31 @@ pub(crate) fn untrimmable_columns(cfg: &ConfigSet) -> UntrimmableColumns {
+        BlobTraceUpdates::Row(updates) => itertools::Either::Left(updates.into_iter()),
+        BlobTraceUpdates::Both((codec, _structured)) => {
+            // This is super unexpected, but not worthy of a panic.
+            tracing::error!("re-encoding structured data?");
```
should we do something here to make this fail in CI?

ParkMyCar replied:
done

danhhz commented on src/persist-client/src/internal/metrics.rs:
```diff
@@ -1073,6 +1102,30 @@ impl CodecMetrics {
 }
 
+#[derive(Debug)]
+pub struct StructuredColumnarMetrics {
```
Can we sneak in a quick metric to cover how much time we spend doing validation plz?

ParkMyCar replied:
I added a metric that tracks decoding time, which is the vast majority of the time spent in validation. After we decode all we do is check if the two values are equal. More than happy to add more metrics if need be!

danhhz commented on src/persist/src/indexed/columnar/parquet.rs:
```diff
@@ -124,25 +137,69 @@ fn encode_parquet_kvtd<W: Write>(
+            // The first 4 columns are our primary (K, V, T, D) and optionally
+            // we also have K_S and/or V_S if we wrote structured data.
+            //
+            // TODO(parkmycar): Add some more validation here.
```
If you've got something in mind, maybe include more detail in the comment

danhhz replied:
Ping

ParkMyCar replied:
Whoops! I just went ahead and added the validation that I could think of

danhhz commented on src/persist-client/src/batch.rs:
```diff
@@ -994,22 +1005,51 @@ impl<T: Timestamp + Codec64> BatchParts<T> {
+                // Only encode our updates in a structured format if required, it's expensive.
+                let s = if cfg.stats_collection_enabled || cfg.batch_columnar_format.is_structured()
+                {
+                    // TODO(parkmycar): Add metrics to track how long columnar encoding takes.
```
I'd vote to do this metrics TODO now

ParkMyCar replied:
done!

danhhz reviewed (approved):
LGTM! left a few comments that would be nice to include in the initial merge of this, especially given that there's still a CI failure to shake out (right?)

would also be worthwhile to think about the various metrics we might want for this when we initially roll it out to staging, etc
- at the very least, I'd love to see timing coverage for the various encode/decode/validate things
- probably also some amount of counts of the major paths, so we can see the dyncfgs taking effect (and that we're not hitting some unexpected more-than-one-row-group case, causing us to fall back to the legacy behavior)
- would be cool if we could get some sense of the size overhead, but not sure if there's a good way to do that without encoding the parquet representation twice. probably not that interesting until we do some non-Plain compression anyway. we do have`mz_persist_shard_usage_current_state_batches_bytes` which will at least let us eyeball how much this goes up when we turn it on

danhhz reviewed (commented):
Ugh, sorry to pile on one more thing, but I was thinking last night and realized we probably want to be able to tell from metadata which batches have the old format and which have the new format. Might be worth doing that from the start? I'd certainly do it before we turn it on more widely in staging, if not in this PR

(maybe ensure that each batch is all one or the other so we don't have to write it down per part?)

ParkMyCar commented:
moving this to draft while I wait for https://github.com/MaterializeInc/materialize/pull/27009 to merge

ParkMyCar commented:
> I'd love to see timing coverage for the various encode/decode/validate things

Added! Currently the metrics are a little wonky because we track the amount of time it takes to encode an entire `Part`, but when decoding we track how long it take to decode a single item. I can push the encoding measurement down to the individual row, let me know!

> probably also some amount of counts of the major paths, so we can see the dyncfgs taking effect (and that we're not hitting some unexpected more-than-one-row-group case, causing us to fall back to the legacy behavior)

Added counts for encoding a Part, decoding a Row, validation, and number of row groups in Parquet.

> would be cool if we could get some sense of the size overhead, but not sure if there's a good way to do that without encoding the parquet representation twice. probably not that interesting until we do some non-Plain compression anyway.

We can get this info from the `FileMetaData` returned from Parquet, added size metrics for all of our top-level columns

ParkMyCar commented:
> Ugh, sorry to pile on one more thing, but I was thinking last night and realized we probably want to be able to tell from metadata which batches have the old format and which have the new format. Might be worth doing that from the start? I'd certainly do it before we turn it on more widely in staging, if not in this PR
> 
> (maybe ensure that each batch is all one or the other so we don't have to write it down per part?)

No worries! Do you mind if I make this change in a follow up PR? This one has already been open long enough that I would like to get it merged. Definitely won't turn this on in Staging (other than my own environment possibly) until making this change though!

ParkMyCar commented:
@bkirwi @danhhz this should be ready for a re-review when you get the chance! Apologies for having force pushed and thus losing the previous reviews progress. The core logic didn't change, other than re-writing the Arrow/Parquet bits with `arrow-rs` and `parquet-rs`, and the majority (all?) of the new changes are around metrics and testing.

ParkMyCar commented on src/persist-types/src/codec_impls.rs:
```diff
@@ -629,10 +629,6 @@ impl ColumnRef<()> for BooleanBuffer {
             .as_any()
             .downcast_ref::<BooleanArray>()
             .ok_or_else(|| format!("expected BooleanArray but was {:?}", array.data_type()))?;
-        if array.logical_nulls().is_some() {
-            return Err("unexpected validity for non-optional bool".to_owned());
-        }
```
I had to remove this because when we encode `SourceData(Err(...))` we push a `Null` into a bool column. So even if the `ColumnType` specifies `nullable: false` it's still possible for the inner column to contain `null` values.

ParkMyCar replied:
The impl of `ColumnRef` for `BooleanBuffer` is the only impl that has an assert like this. I don't know for sure but my guess is that when roundtripping through Parquet the boolean column gets the validity buffer of the higher level `StructArray` which is how this assert ends up failing

ParkMyCar commented on src/storage-types/src/sources.rs:
```diff
@@ -1290,11 +1292,16 @@ mod tests {
+        scalar_type: ScalarType,
+        datums: impl IntoIterator<Item = Datum<'a>>,
+    ) {
+        use mz_persist_types::parquet::validate_roundtrip;
```
Using the parquet encoding here is what caught the bug in `arrow2` that motivated the switch to `arrow-rs`, previously we were only round-tripping through a `Part` which doesn't exercise structured Arrow or Parquet

benesch commented on test/testdrive/mzcompose.py:
```diff
@@ -90,9 +97,23 @@ def workflow_default(c: Composition, parser: WorkflowArgumentParser) -> None:
+
+    additional_system_parameter_defaults = {}
+    for val in sysparams:
+        x = val[0].split("=")
```
You only one to split once; it's legal for the parameter value to contain `=`.

```suggestion
        x = val[0].split("=", maxsplit=1)
```

ParkMyCar replied:
Done!

benesch reviewed (commented):
Changes to `ore::assert` LGTM. I'm unsubscribing to reduce email noise, so tag me explicitly if you need me back.

danhhz commented on src/persist-client/src/fetch.rs:
```diff
@@ -796,6 +896,35 @@ where
+                    // Purposefully do not trace to prevent blowing up Sentry.
+                    let is_valid = val_metrics.report_valid(|| Ok(v_s) == v);
+                    if !is_valid {
+                        soft_panic_no_log!("structured key did not match");
```
nit: key->val

ParkMyCar replied:
done!

danhhz commented on src/storage-types/src/sources.rs:
```diff
@@ -1317,7 +1324,23 @@ mod tests {
     fn all_scalar_types_columnar_roundtrip() {
```
nit: Should this and the other test name be s/columnar/parquet/ now that we're using `parquet::validate_roundtrip` instead of `columnar::validate_roundtrip`?

ParkMyCar replied:
Good idea, updated!

danhhz commented on src/persist-types/src/dyn_struct.rs:
```diff
@@ -218,17 +218,25 @@ impl DynStructCol {
         let (mut fields, mut arrays) = (Vec::new(), Vec::new());
         for (name, _stats_fn, col) in self.cols() {
             let (array, is_nullable) = col.to_arrow();
+            if array.len() == 0 && self.len != 0 {
```
Can we stick a comment here? (Also this seems slightly wrong/odd/surprising, but maybe the comment will enlighten me)

ParkMyCar replied:
Shoot, sorry I left this in, it was a remnant of when we used to map a `StructArray` with no fields to an empty `NullArray`. Because we changed that to be a NullArray with a length, we no longer need this check. Removed!

danhhz commented on src/persist-client/src/batch.rs:
```diff
@@ -1022,22 +1033,54 @@ impl<T: Timestamp + Codec64> BatchParts<T> {
+                let stats = if let Some((part, stats)) = s {
+                    // Only change the updates to use the structured format, if our updates had a
+                    // single batch.
+                    match updates.updates {
```
No concrete counter-proposal, but I'm having trouble following the control flow of this if/else structure, both exactly what happens in the various conditions and why. For example, feels like this could be behind only `cfg.batch_columnar_format.is_structured()` but it's also gated by `cfg.stats_collection_enabled` (or maybe I'm reading the code wrong! not confidant)

ParkMyCar replied:
Yeah the flow isn't super clear, the tricky bit is if we somehow fail to encode in our columnar format we don't (yet?) want to panic and instead bail on stats collection. I re-worked this using labeled blocks which I think allows it to be a bit cleaner, let me know what you think!

danhhz reviewed (commented):
> > Ugh, sorry to pile on one more thing, but I was thinking last night and realized we probably want to be able to tell from metadata which batches have the old format and which have the new format. Might be worth doing that from the start? I'd certainly do it before we turn it on more widely in staging, if not in this PR
> > (maybe ensure that each batch is all one or the other so we don't have to write it down per part?)
> 
> No worries! Do you mind if I make this change in a follow up PR? This one has already been open long enough that I would like to get it merged. Definitely won't turn this on in Staging (other than my own environment possibly) until making this change though!

Yeah, 100% fine if we're not turning it on for other envs on staging. You should definitely try it on yours/mine as-is, but then let's make sure to wipe them after so we don't confuse ourselves later.

> Apologies for having force pushed and thus losing the previous reviews progress.
Not at all! I can't imagine trying to do that rebase without a force push

bkirwi commented on src/persist/src/indexed/columnar/parquet.rs:
```diff
@@ -24,29 +26,30 @@ use parquet::file::properties::{EnabledStatistics, WriterProperties, WriterVersi
 const INLINE_METADATA_KEY: &str = "MZ:inline";
 
-/// Encodes an BlobTraceBatchPart into the Parquet format.
+/// Encodes a [`BlobTraceBatchPart`] into the Parquet format.
```
🙇

bkirwi commented on src/storage-types/src/sources.rs:
```diff
@@ -1199,8 +1201,8 @@ impl PartDecoder<SourceData> for SourceDataDecoder {
-            (true, Some(_)) | (false, None) => {
-                panic!("SourceData should have exactly one of ok or err")
+            x @ (true, Some(_)) | x @ (false, None) => {
+                panic!("SourceData should have exactly one of ok or err, {idx} @ {x:?}\n{self:?}")
```
We may want to crank this back slightly... IIRC we don't want user-level errors in our logs.

ParkMyCar replied:
Makes sense! I added this originally when doing some of the debugging related to `arrow2`, removed

ParkMyCar commented on src/persist/src/metrics.rs:
```diff
@@ -78,22 +80,181 @@ impl S3BlobMetrics {
+            var_labels: ["op", "column"],
+        ));
+
+        let part_build_duration: CounterVec = registry.register(metric!(
```
Ah ha, they definitely don't need to be, fixed!

ParkMyCar commented on src/persist/src/indexed/encoding.rs:
```diff
@@ -94,7 +152,61 @@ pub struct BlobTraceBatchPart<T> {
+    /// an Apache Arrow columnar format.
+    ///
+    /// [`Codec`]: mz_persist_types::Codec
+    Both((ColumnarRecords, ColumnarRecordsStructuredExt)),
```
Good thought, reworked this like you suggested

danhhz commented:
\o/

ParkMyCar commented:
Merged to make this week's release, more than happy to follow up with any requested changes :)

danhhz commented:
None from me, lgtm!
